_posts/2025-11-07-introducing-kubeflow-sdk.md

```diff
@@ -0,0 +1,374 @@
+
+The client works with pre-configured runtimes that Platform Administrators set up. These runtimes define the container images, resource policies, and infrastructure settings. As an AI Practitioner, you reference these runtimes and focus on your training code:
+
+```python
+from kubeflow.trainer import TrainerClient, CustomTrainer
+
+def get_torch_dist():
+    """Your PyTorch training code runs on each node."""
+    import os
+    import torch
+    import torch.distributed as dist
+
+    dist.init_process_group(backend="gloo")
+    print("PyTorch Distributed Environment")
+    print(f"WORLD_SIZE: {dist.get_world_size()}")
+    print(f"RANK: {dist.get_rank()}")
+    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")
+
+# Create the TrainJob
+job_id = TrainerClient().train(
+    runtime=TrainerClient().get_runtime("torch-distributed"),
+    trainer=CustomTrainer(
+        func=get_torch_dist,
+        num_nodes=3,
+        resources_per_node={
+            "cpu": 2,
+        },
+    ),
+)
+
+# Wait for TrainJob to complete
+TrainerClient().wait_for_job_status(job_id)
+
+# Print TrainJob logs
+print("\n".join(TrainerClient().get_job_logs(name=job_id)))
+```
+
+The TrainerClient supports `CustomTrainer` for your own training logic and [`BuiltinTrainer`](https://www.kubeflow.org/docs/components/trainer/user-guides/builtin-trainer/torchtune/) for pre-packaged training patterns like LLM fine-tuning. For example, fine-tuning LLMs with TorchTune:
+
+```python
+from kubeflow.trainer import TrainerClient, BuiltinTrainer, TorchTuneConfig
+from kubeflow.trainer import Initializer, HuggingFaceDatasetInitializer, HuggingFaceModelInitializer
+from kubeflow.trainer import TorchTuneInstructDataset, LoraConfig, DataFormat
+
+client = TrainerClient()
+
+client.train(
+    runtime=client.get_runtime(name="torchtune-llama3.2-1b"),
+    initializer=Initializer(
+        dataset=HuggingFaceDatasetInitializer(
+            storage_uri="hf://tatsu-lab/alpaca/data"
+        ),
+        model=HuggingFaceModelInitializer(
+            storage_uri="hf://meta-llama/Llama-3.2-1B-Instruct",
+            access_token="hf_...",
+        )
+    ),
+    trainer=BuiltinTrainer(
+        config=TorchTuneConfig(
+            dataset_preprocess_config=TorchTuneInstructDataset(
+                source=DataFormat.PARQUET,
+            ),
+            peft_config=LoraConfig(
+                apply_lora_to_mlp=True,
+                lora_attn_modules=["q_proj", "k_proj", "v_proj", "output_proj"],
+                quantize_base=True,
+            ),
+            resources_per_node={
+                "gpu": 1,
+            }
+        )
+    )
+)
+```
+
+One click fine tuning can also be accomplished without any configuration being set. The default model, dataset and training configurations are pre-baked into the runtime:
+
+```python
+TrainerClient().train(
+    runtime=TrainerClient().get_runtime("torchtune-qwen2.5-1.5b"),
```

Comment:
We may still want to show how to pass the dataset users would want to fine-tune the model with?